### PR TITLE
Merklized but not diffed

### PR DESCRIPTION
**Pull request by yito88**
## Describe your changes
- Update some data in the subspace and the merkle tree, but not update the diffs
    - Add `diff_key_filter` to check if the data should be stored to diffs even if `merkle_tree_key_filter` is true
- Add `NoDiff` subtree in Merkle tree
    - Need it because we can't rebuild each subtree when some data aren't in diffs

## Indicate on which release or other PRs this topic is based on
`v0.36.1`

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

**Review by batconjurer — DISMISSED** [review]

File: crates/node/src/shell/mod.rs [hunk 1]
```diff
@@ -355,7 +355,14 @@ where
 
 /// Merkle tree storage key filter. Return `false` for keys that shouldn't be
 /// merklized.
-pub fn is_merklized_storage_key(key: &namada_sdk::storage::Key) -> bool {
+pub fn is_merklized_storage_key(_key: &namada_sdk::storage::Key) -> bool {
```
Comment (cwgoes) [inline crates/node/src/shell/mod.rs:358]: We should just get rid of this function, honestly.

File: crates/node/src/storage/mod.rs [hunk 2]
```diff
@@ -544,6 +563,41 @@ mod tests {
...
+            if key != client_counter_key() {
+                continue;
+            }
+            let merkle_key =
+                Key::from(NO_DIFF_KEY_PREFIX.to_string().to_db_key())
+                    .join(&key);
+            let tree =
+                state.get_merkle_tree(height, Some(StoreType::NoDiff))?;
+            assert_eq!(tree.root().0, roots.get(&height).unwrap().0);
+            match write_type {
```
Comment (cwgoes) [inline crates/node/src/storage/mod.rs:582]: Needs comments, this logic is not clear.

File: crates/node/src/shell/mod.rs (hunk 1 above)
Comment (yito88) [reply crates/node/src/shell/mod.rs]: Removed cd441a2

File: crates/node/src/storage/mod.rs (hunk 2 above)
Comment (yito88) [reply crates/node/src/storage/mod.rs:580]: added a comment cd441a2

File: crates/node/src/storage/mod.rs [hunk 3]
```diff
@@ -437,15 +445,23 @@ mod tests {
             address::testing::nam(),
             None,
             is_merklized_storage_key,
+            is_key_diff_storable,
         );
+        let make_key = |suffix: u64| {
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply crates/node/src/storage/mod.rs:445]: Can you explain what this whole `suffix % 3u64` logic is doing, though?

File: crates/node/src/storage/mod.rs (hunk 2 above)
Comment (cwgoes) [reply crates/node/src/storage/mod.rs:580]: Can you explain the semantics of these 0/1/3 write types?

**Review by cwgoes — CHANGES_REQUESTED** [review]
I still want clearer comments 😁

File: crates/node/src/storage/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (yito88) [reply crates/node/src/storage/mod.rs:445]: 44c9201

File: crates/node/src/storage/mod.rs (hunk 2 above)
Comment (yito88) [reply crates/node/src/storage/mod.rs:586]: 44c9201

File: crates/node/src/storage/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (brentstone) [reply crates/node/src/storage/mod.rs:445]: Also what does "probable" mean in this context?

**Comment by brentstone** [comment]
@yito88 @cwgoes do we need to use this for replay protection too or do we only need what is already in #3284?

File: crates/node/src/storage/mod.rs [hunk 4]
```diff
@@ -513,37 +527,84 @@ mod tests {
...
+            match write_type {
+                0 => {
+                    // data was not updated
+                    if *current_state.get(&key).unwrap() {
+                        assert!(tree.has_key(&merkle_key)?);
+                    } else {
+                        assert!(!tree.has_key(&merkle_key)?);
+                    }
+                }
+                1 | 3 => {
```
Comment (cwgoes) [inline crates/node/src/storage/mod.rs:595]: but what do these numbers 1 or 3 mean?

**Review by cwgoes — CHANGES_REQUESTED** [review]

**Comment by cwgoes** [comment]
> @yito88 @cwgoes do we need to use this for replay protection too or do we only need what is already in #3284?

We do not need to use this for replay protection, but we do need to merge this to include everything in the app hash.

File: crates/node/src/storage/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (yito88) [reply crates/node/src/storage/mod.rs:445]: these data require proofs of the existence or non-existence

**Review by cwgoes — APPROVED** [review]

**Pull request merged**